yarn-project/ethereum/src/l1_tx_utils.ts

```
@@ -777,7 +779,7 @@ export class L1TxUtils extends ReadOnlyL1TxUtils {
 
         const currentNonce = await this.client.getTransactionCount({ address: account });
         if (currentNonce > nonce) {
-          for (const hash of txHashes) {
+          for (const hash of allVersions) {
```

Should we set state to `MINED` and throw if we looped over `allVersions` and not found the receipt? That would have fixed the issue as well, and will also catch any instances where we fail to populate `allVersions` with all possible hashes (eg node going down).

---

Yes, anything that adds resiliency to this I think is a good idea.

---

Though, doesn't this `const receipt = await this.client.getTransactionReceipt({ hash });` only return a receipt for mined transactions. If we don't find one wouldn't that just mean nothing is yet mined?

---

If we don't find a receipt after looping over all hashes in `allVersions`, but `currentNonce > nonce`, it means that _some_ tx from our sender with the same nonce was mined, but none that we were tracking. And none of the ones we were tracking will get mined, since another one with that nonce made it.

FWIW some libraries (`viem` included when you `waitForTxReceipt`) will actually download all mined tx hashes for every single block and check sender/nonce, to see if your tx was replaced and by what. This is expensive in terms of rpc usage but required in UIs because wallets do not report speed-up txs back to the app, so the app has [no other way to detect the replacement.](https://github.com/ethereum/execution-apis/issues/494) But in scenarios where we control every tx that is sent like this one, we shouldn't need this, barring some bug in our code.

---

Ah yes, I missed the nonce check!